lessons/gitBranches/lesson.md

```diff
@@ -0,0 +1,162 @@
+messing with the master dataset. After you confirm the problem is real and that
+you have the solution, you submit a pull request of the `fixing` branch to be
+merged with the `master` branch.
+- What is often the case in software development, a bug or missing feature in
+the software gets identified. Because the software is already in production use
+(fairly stable, other people rely on it, etc), you can't just make changes to
+the main software code. So a `hotfix` or `feature` branch is created to address
+these problems, which will eventually get merged in with the `master` branch for
+the next version of the software. This ensures that other people's code isn't
+broken everytime a bug gets fixed.
+
+There are many uses of branches in Git. The nice (and very powerful) thing about
+Git is the fact that branches are very cheap compared to other version control
+systems. By cheap, I mean they don't take up much disk space, it's
+computationally easy to move between branches, and it's (relatively) easy to
+merge branches together. This is because of how Git represents branches, since
+they are simply *pointers* or an individual commit. *That's it.* Just a
+pointer... Git commit history is a 
+[directed acyclic graph](https://en.wikipedia.org/wiki/Directed_acyclic_graph), 
+which means that every single commit always has a 'parent' commit (the
+previous commit in the history, or multiple parents when a merge happens), and
+any individual commit can have multiple 'children'. This history can be traced
+back through the 'lineage' or 'ancestry'. The branch just gives a name to each
+'lineage' when a commit has multiple children.
+
+When you merge two branches together, the commit histories get merged together
+as well. Which means that all the changes you made in each branch gets combined
+back into a single lineage, rather than two. This makes it easier to work
+collaboratively on a project, since each individual could work on their own
+branches, without dealing with the messiness that could come from working all on
+one branch.
+
+## Commands used with branches
+
+Branches are best understood visually. So let's first start with using 
+[this website](https://onlywei.github.io/explain-git-with-d3/) to see what the 
+`branch`, `checkout`, and `merge` commands are doing.
+
+After we've tried that, let's do it locally (on your own computer). Here is a
+sequence of commands to try out:
+
+    cd ~/Desktop
+    mkdir git-branches
+    cd git-branches
+    git init # start a repo
+    git add .
+    git commit -m "First commit" # make the first commit
+    git branch testBranch
+    git checkout testBranch # create and move to a new branch
+    ## can also do git checkout -b testBranch
+    echo "Some text" > file.txt 
+    git add file.txt
+    git commit -m "Added a file with text"
+    git checkout master
+    echo "Text in another file" > new-file.txt
+    git add new-file.txt
+    git commit -m "Added another file"
+    git log --graph --oneline
```

Comment:
You might also want `--decorate` to see the branch names and `--all` to list all branches.